symengine/logic.cpp

```diff
@@ -146,4 +146,186 @@ int Piecewise::compare(const Basic &o) const
+}
+
+int Or::compare(const Basic &o) const
+{
+    SYMENGINE_ASSERT(is_a<Or>(o))
+    return unified_compare(args, static_cast<const Or &>(o).args);
+}
+
+bool Or::is_canonical(const set_boolean &container)
+{
+    if (container.size() >= 2) {
+        for (auto &a : container) {
+            if (is_a<BooleanAtom>(*a) or is_a<Or>(*a))
+                return false;
+            if (container.find(logical_not(a)) != container.end())
+                return false;
+        }
+        return true;
+    }
+    return false;
+}
+
+Not::Not(const RCP<const Boolean> &in) : arg{in}
+{
+    SYMENGINE_ASSERT(is_canonical(in));
+}
+
+std::size_t Not::__hash__() const
+{
+    std::size_t seed = NOT;
+    hash_combine<Basic>(seed, *arg);
+    return seed;
+}
+
+vec_basic Not::get_args() const
+{
+    vec_basic v;
+    v.push_back(arg);
+    return v;
+}
+
+bool Not::__eq__(const Basic &o) const
+{
+    return is_a<Not>(o) and eq(*arg, *static_cast<const Not &>(o).arg);
+}
+
+int Not::compare(const Basic &o) const
+{
+    SYMENGINE_ASSERT(is_a<Not>(o))
+    return arg->__cmp__(*static_cast<const Not &>(o).arg);
+}
+
+bool Not::is_canonical(const RCP<const Boolean> &in)
+{
+    if (is_a<BooleanAtom>(*in) or is_a<Not>(*in))
+        return false;
+    return true;
+}
+
+template <typename caller>
+RCP<const Boolean> and_or(const set_boolean &s, const bool &op_x_notx)
+{
+    set_boolean args;
+    for (auto &a : s) {
+        if (is_a<BooleanAtom>(*a)) {
+            auto val = static_cast<const BooleanAtom &>(*a).get_val();
+            if (val == op_x_notx)
+                return boolean(op_x_notx);
+            else
+                continue;
+        }
+        if (is_a<caller>(*a)) {
+            const caller &to_insert = static_cast<const caller &>(*a);
+            args.insert(to_insert.args.begin(), to_insert.args.end());
+            continue;
+        }
+        args.insert(a);
+    }
+    for (auto &a : args) {
+        if (args.find(logical_not(a)) != args.end())
+            return boolean(op_x_notx);
+    }
+    if (args.size() == 1)
+        return *(args.begin());
+    else if (args.size() == 0)
+        return boolean(not op_x_notx);
+    return make_rcp<const caller>(args);
+}
+
+RCP<const Boolean> logical_not(const RCP<const Boolean> &s)
+{
+    if (is_a<BooleanAtom>(*s)) {
+        const BooleanAtom &a = static_cast<const BooleanAtom &>(*s);
+        return boolean(not a.get_val());
+    } else if (is_a<Not>(*s)) {
+        const Not &a = static_cast<const Not &>(*s);
+        return a.arg;
+    } else if (is_a<Or>(*s)) {
+        const Or &o = static_cast<const Or &>(*s);
+        set_boolean s;
+        for (auto &a : o.args) {
+            s.insert(logical_not(a));
+        }
+        return logical_and(s);
+    } else if (is_a<And>(*s)) {
+        const And &o = static_cast<const And &>(*s);
+        set_boolean s;
+        for (auto &a : o.args) {
+            s.insert(logical_not(a));
+        }
+        return logical_or(s);
```

Comment:
What do you mean by, `s might be a Contains object.`? `s` is a `set_boolean` right?